### DISCUSSION TITLE
Reducing cross-AZ costs / WAN usage

### DISCUSSION THREAD

**Discussion by jimmyjones2** [rook/rook#11929] [Q&A]
Hey,

I've got separate clusters in pubic cloud (stretched over multiple AZ) and on-prem (stretched over multiple datacenters). For the former, I want to keep cross-AZ traffic to a minimum due to the cost, and for on-prem to keep the network people happy and for better performance (our WAN links aren't that fast).

I've seen in v1.11 I can enable [RBD read affinity](https://github.com/rook/rook/pull/11631), which I think will solve the issue when reading from my RWO PVCs? Will there be future support for write affinity?

Is there a similar feature available or planned for cephfs / RWX PVCs?

For object storage, in CephObjectStore I can create multiple gateway pods spread over different AZ/datacenters, but do they support read affinity when reading OSDs plus I believe it would also be necessary to add [topology aware hints](https://kubernetes.io/docs/concepts/services-networking/topology-aware-hints/) on the gateway Service? Otherwise worst case my app could call a gateway pod running in another AZ/datacenter which in turn could access the OSD in the original AZ/datacenter - 2 network hops, when it could have been 0! Are there future features planned for write affinity - ideally a PUT would write to an local OSD, then replicate to the other AZ/datacenters - rather than having a chance of writing to a remote AZ/datacenter then replicating back to the local one.

Thanks :)

**Comment by BlaineEXE** [comment]
Write affinity would reduce the failure tolerance of the Ceph cluster because all data would be limited to one zone, so it isn't likely to be a feature added to Ceph. Any data written has to be replicated to all failure zones for a write to be guaranteed.

**Reply by jimmyjones2** [reply]
I just started drawing a diagram of a 3 datacenter/AZ setup and think I've found my misunderstanding - when a pod writes to ceph storage spread over 3 datacenters, the ceph client will open a connection and write the same data to all 3, so regardless of where the primary is, there will always be exactly two WAN/cross-AZ connections? (I was assuming the client only wrote to the primary OSD, which then initiated the replication to the other 2)

How about read affinity for cephfs and for object store (plus topology aware hints) to always serve read requests from the same datacenter and avoid WAN/cross-AZ connections?
